File: android/app/src/main/java/io/bluewallet/bluewallet/BitcoinPriceWidget.java
package io.bluewallet.bluewallet;

import android.app.Application;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.PowerManager;
import android.util.Log;
import android.widget.RemoteViews;

import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import java.text.NumberFormat;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

public class BitcoinPriceWidget extends AppWidgetProvider {

    private static final String TAG = "BitcoinPriceWidget";
    private static final String ACTION_UPDATE = "io.bluewallet.bluewallet.UPDATE_WIDGET";
    private static final long UPDATE_INTERVAL_MINUTES = 15; // Update interval in minutes
    private static final int MAX_RETRIES = 3;

    private static PowerManager.WakeLock wakeLock;
    private static int retryCount = 0;
    private static boolean isScreenOn = true;

    @Override
    public void onEnabled(Context context) {
        super.onEnabled(context);
        registerScreenReceiver(context);
        schedulePeriodicUpdates(context);
    }

    @Override
    public void onDisabled(Context context) {
        super.onDisabled(context);
        unregisterScreenReceiver(context);
        WorkManager.getInstance(context).cancelUniqueWork("UpdateWidgetWork");
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);
        if (isScreenOn) {
            scheduleWork(context);
        }
    }

    private void schedulePeriodicUpdates(Context context) {
        PeriodicWorkRequest workRequest = new PeriodicWorkRequest.Builder(WidgetUpdateWorker.class,
                UPDATE_INTERVAL_MINUTES, TimeUnit.MINUTES)
                .setInitialDelay(UPDATE_INTERVAL_MINUTES, TimeUnit.MINUTES)
                .build();

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
                "UpdateWidgetWork",
                ExistingPeriodicWorkPolicy.REPLACE,
                workRequest
        );
    }

    private void registerScreenReceiver(Context context) {
        IntentFilter filter = new IntentFilter();
        filter.addAction(Intent.ACTION_SCREEN_ON);
        filter.addAction(Intent.ACTION_SCREEN_OFF);
        context.getApplicationContext().registerReceiver(screenReceiver, filter);
    }

    private void unregisterScreenReceiver(Context context) {
        context.getApplicationContext().unregisterReceiver(screenReceiver);
    }

    private final BroadcastReceiver screenReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction() != null) {
                PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
                if (pm != null) {
                    switch (intent.getAction()) {
                        case Intent.ACTION_SCREEN_ON:
                            isScreenOn = true;
                            Log.d(TAG, "Screen ON");
                            acquireWakeLock(context);
                            scheduleWork(context);
                            break;
                        case Intent.ACTION_SCREEN_OFF:
                            isScreenOn = false;
                            Log.d(TAG, "Screen OFF");
                            releaseWakeLock();
                            break;
                    }
                }
            }
        }
    };

    private void acquireWakeLock(Context context) {
        if (wakeLock == null) {
            PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
            if (pm != null) {
                wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
                wakeLock.acquire(10 * 60 * 1000L /*10 minutes*/);
            }
        }
    }

    private void releaseWakeLock() {
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
            wakeLock = null;
        }
    }

    private void scheduleWork(Context context) {
        Log.d(TAG, "Scheduling work for widget update");
        WorkManager.getInstance(context).enqueue(WidgetUpdateWorker.createWorkRequest());
    }
}

File: android/app/src/main/java/io/bluewallet/bluewallet/BitcoinPriceWidget.kt
package io.bluewallet.bluewallet

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.RemoteViews
import androidx.work.WorkManager

class BitcoinPriceWidget : AppWidgetProvider() {

    companion object {
        private const val TAG = "BitcoinPriceWidget"
        private const val SHARED_PREF_NAME = "group.io.bluewallet.bluewallet"
        
        fun updateNetworkStatus(context: Context, appWidgetIds: IntArray) {
            val isNetworkAvailable = NetworkUtils.isNetworkAvailable(context)
            val appWidgetManager = AppWidgetManager.getInstance(context)
            
            for (appWidgetId in appWidgetIds) {
                val views = RemoteViews(context.packageName, R.layout.widget_layout)
                views.setViewVisibility(R.id.network_status, if (isNetworkAvailable) View.GONE else View.VISIBLE)
                appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views)
            }
        }

        fun refreshWidget(context: Context, appWidgetId: Int) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            
            // Create new RemoteViews to ensure it picks up current theme
            val views = RemoteViews(context.packageName, R.layout.widget_layout)
            
            // Set network status
            val isNetworkAvailable = NetworkUtils.isNetworkAvailable(context)
            views.setViewVisibility(R.id.network_status, if (isNetworkAvailable) View.GONE else View.VISIBLE)
            
            // Try to load cached data first
            val sharedPref = context.getSharedPreferences(SHARED_PREF_NAME, Context.MODE_PRIVATE)
            val cachedPrice = sharedPref.getString("previous_price", null)
            
            if (cachedPrice != null) {
                // Show cached data immediately
                val preferredCurrency = sharedPref.getString("preferredCurrency", "USD")
                val preferredCurrencyLocale = sharedPref.getString("preferredCurrencyLocale", "en-US")
                
                try {
                    val localeParts = preferredCurrencyLocale?.split("-") ?: listOf("en", "US")
                    val locale = if (localeParts.size == 2) {
                        java.util.Locale(localeParts[0], localeParts[1])
                    } else {
                        java.util.Locale.getDefault()
                    }
                    val currencyFormat = java.text.NumberFormat.getCurrencyInstance(locale)
                    val currency = java.util.Currency.getInstance(preferredCurrency ?: "USD")
                    currencyFormat.currency = currency
                    currencyFormat.maximumFractionDigits = 0
                    
                    views.setViewVisibility(R.id.loading_indicator, View.GONE)
                    views.setViewVisibility(R.id.price_value, View.VISIBLE)
                    views.setViewVisibility(R.id.last_updated_label, View.VISIBLE)
                    views.setViewVisibility(R.id.last_updated_time, View.VISIBLE)
                    views.setTextViewText(R.id.price_value, currencyFormat.format(cachedPrice.toDouble().toInt()))
                    views.setTextViewText(R.id.last_updated_time, java.text.SimpleDateFormat("hh:mm a", java.util.Locale.getDefault()).format(java.util.Date()))
                    views.setViewVisibility(R.id.price_arrow_container, View.GONE)
                } catch (e: Exception) {
                    Log.e(TAG, "Error displaying cached price", e)
                    // Show loading state if cache display fails
                    views.setViewVisibility(R.id.loading_indicator, View.VISIBLE)
                    views.setViewVisibility(R.id.price_value, View.GONE)
                    views.setViewVisibility(R.id.last_updated_label, View.GONE)
                    views.setViewVisibility(R.id.last_updated_time, View.GONE)
                    views.setViewVisibility(R.id.price_arrow_container, View.GONE)
                }
            } else {
                // No cached data, show loading state
                views.setViewVisibility(R.id.loading_indicator, View.VISIBLE)
                views.setViewVisibility(R.id.price_value, View.GONE)
                views.setViewVisibility(R.id.last_updated_label, View.GONE)
                views.setViewVisibility(R.id.last_updated_time, View.GONE)
                views.setViewVisibility(R.id.price_arrow_container, View.GONE)
            }
            
            appWidgetManager.updateAppWidget(appWidgetId, views)
            WidgetUpdateWorker.scheduleImmediateUpdate(context)
            WidgetUpdateWorker.scheduleWork(context)
        }
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
        
        for (widgetId in appWidgetIds) {
            Log.d(TAG, "Updating widget with ID: $widgetId")
            refreshWidget(context, widgetId)
        }
    }

    override fun onEnabled(context: Context) {
        super.onEnabled(context)
        WidgetUpdateWorker.scheduleImmediateUpdate(context)
        WidgetUpdateWorker.scheduleWork(context)
    }

    override fun onDisabled(context: Context) {
        super.onDisabled(context)
        context.getSharedPreferences(SHARED_PREF_NAME, Context.MODE_PRIVATE).edit().clear().apply()
        WorkManager.getInstance(context).cancelUniqueWork(WidgetUpdateWorker.WORK_NAME)
    }

    override fun onAppWidgetOptionsChanged(context: Context, appWidgetManager: AppWidgetManager, 
                                          appWidgetId: Int, newOptions: Bundle?) {
        super.onAppWidgetOptionsChanged(context, appWidgetManager, appWidgetId, newOptions)
        refreshWidget(context, appWidgetId)
    }
}